### PR TITLE
adding tgn node prop pred example

### PR DESCRIPTION
**Pull request by shenyangHuang**
### Summary / Description

adding node prop pred example 

For CTDG methods on node prop pred, to match the same performance with TGB we shouldn't discretize, because it means less frequent updates potentially. 

Might need to adjust this for DyGformer as well

Validation ndcg=0.3936 for TGN on `tgbn-trade`

### CODE REVIEW

File: tgm/hooks/hooks.py [hunk 1]
```diff
@@ -307,7 +318,8 @@ def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
             batch.nbr_times.append(nbr_times)  # type: ignore
             batch.nbr_feats.append(nbr_feats)  # type: ignore
 
-        self._update(batch)
+        if batch.src.numel():
```
Comment (Jacob-Chmura) [inline tgm/hooks/hooks.py:321]: Nice, creating an issue to add a test case for node-only event batches

File: tgm/hooks/hooks.py [hunk 2]
```diff
@@ -246,6 +247,7 @@ def __init__(
         num_nbrs: List[int],
         edge_feats_dim: int,
         directed: bool = False,
+        seed_nodes_key: str = None,  # type: ignore
```
Comment (Jacob-Chmura) [inline tgm/hooks/hooks.py:250]: Creating an issue to manage this across both uniforma and recency sampler

File: tgm/hooks/hooks.py [hunk 3]
```diff
@@ -285,14 +288,22 @@ def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
...
+                        times.append(batch.neg_time)
+                    seed_nodes = torch.cat(seed)
+                    seed_times = torch.cat(times)
+                else:
+                    seed_nodes = getattr(batch, self._seed_nodes_key)
+                    if seed_nodes is None:
+                        return batch
+                    else:
+                        seed_nodes = seed_nodes.to(device)
+                        seed_times = batch.node_ids.to(device)  # type: ignore #! to adjust later, it should point to node event time
```
Comment (Jacob-Chmura) [inline tgm/hooks/hooks.py]: Can't we just so `seed_times = batch.node_times`?

File: tgm/hooks/hooks.py [hunk 4]
```diff
@@ -285,14 +288,22 @@ def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
...
+                    seed = [batch.src, batch.dst]
+                    times = [batch.time.repeat(2)]  # Real link times
+                    if hasattr(batch, 'neg'):
+                        batch.neg = batch.neg.to(device)
+                        seed.append(batch.neg)
+                        times.append(batch.neg_time)
+                    seed_nodes = torch.cat(seed)
+                    seed_times = torch.cat(times)
+                else:
+                    seed_nodes = getattr(batch, self._seed_nodes_key)
```
Comment (Jacob-Chmura) [inline tgm/hooks/hooks.py:301]: This should probably be a list when we actually do it. I may want seed nodes to be something like `batch.node_ids, batch.dst, batch.foo`

And technically, this should make each of these part of the `requires()` so that for example, if you try to do node sampling (`batch.node_ids`) and then you run on a graph without dynamic node events, you get a reasonable error saying this hook cannot be resolved on a graph that has empty dynamic node features

File: examples/nodeproppred/tgn.py [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+    msg_dim=test_dg.edge_feats_dim,
+    time_enc=memory.time_enc,
+).to(args.device)
+decoder = NodePredictor(in_dim=args.embed_dim, out_dim=num_classes).to(args.device)
+opt = torch.optim.Adam(
+    set(memory.parameters()) | set(encoder.parameters()) | set(decoder.parameters()),
+    lr=args.lr,
+)
+
+with Usage(gpu=True):
```
Comment (Jacob-Chmura) [inline examples/nodeproppred/tgn.py]: We should remove this before merging.

Comment (Jacob-Chmura) [reply examples/nodeproppred/tgn.py]: Also, I created an issue, that our exampels should manager measurement and profiling better #251

File: tgm/hooks/hooks.py (hunk 3 above)
Comment (shenyangHuang) [reply tgm/hooks/hooks.py]: yes, this is actually a bug I think

Comment (shenyangHuang) [reply tgm/hooks/hooks.py]: fixed

File: examples/nodeproppred/tgn.py (hunk 5 above)
Comment (shenyangHuang) [reply examples/nodeproppred/tgn.py]: removed

File: tgm/hooks/hooks.py (hunk 4 above)
Comment (shenyangHuang) [reply tgm/hooks/hooks.py:301]: I agree, but this should be a systematic PR

**Review by Jacob-Chmura — APPROVED** [review]
